trajectory_executor/src/trajectory_executor/trajectory_executor.cpp

```diff
@@ -0,0 +1,188 @@
+                    description_builder << "No match found for control plugin " 
+                        << control_plugin << " at point " 
+                        << _timesteps_since_last_traj << " in current trajectory!";
+                    shutdown(description_builder.str());
+                }
+                _timesteps_since_last_traj++;
+            } else {
+                shutdown("Ran out of trajectory data to consume!");
+            }
+        } else {
+            ROS_DEBUG("Awaiting initial trajectory publication...");
+        }
+        ROS_DEBUG("TrajectoryExecutor tick completed succesfully!");
+    }
+
+    void TrajectoryExecutor::shutdown(std::string reason) {
+        ROS_ERROR_STREAM("FATAL ERROR: " << reason);
+        cav_msgs::SystemAlert system_alert;
+        system_alert.type = system_alert.FATAL;
+        system_alert.description = reason;
+        ROS_FATAL("Publishing FATAL error on system alert topic!");
+        _default_nh->publishSystemAlert(system_alert);
+        ros::shutdown();
+
+        exit(1);
+    }
+
+    void TrajectoryExecutor::run()
+    {
+        ROS_DEBUG("Starting operations for TrajectoryExecutor component...");
+        _timer = _timer_nh->createTimer(
+            ros::Duration(ros::Rate(this->_min_traj_publish_tickrate_hz)),
+            &TrajectoryExecutor::onTrajEmitTick, 
+            this);
+
+        ros::AsyncSpinner timer_spinner(1, &_timer_callbacks);
+        timer_spinner.start();
+        ROS_DEBUG("Timer thread started!");
+
+        ros::AsyncSpinner msg_spinner(1, &_msg_callbacks);
+        msg_spinner.start();
+
+        ROS_DEBUG("TrajectoryExecutor component started succesfully! Starting to spin.");
+
+        ros::Rate r(_default_spin_rate);
+        while (ros::ok())
+        {
+            ros::spinOnce();
+            r.sleep();
+        }
+
+        timer_spinner.stop();
+        msg_spinner.stop();
+
+        ros::shutdown();
+    }
+
+    bool TrajectoryExecutor::init()
+    {
+        ROS_DEBUG("Initializing TrajectoryExecutor node...");
+    
+        _default_nh = std::unique_ptr<ros::CARMANodeHandle>(new ros::CARMANodeHandle);
+        _timer_nh = std::unique_ptr<ros::CARMANodeHandle>(new ros::CARMANodeHandle);
+        _timer_nh->setCallbackQueue(&_timer_callbacks);
+        _msg_nh = std::unique_ptr<ros::CARMANodeHandle>(new ros::CARMANodeHandle);
+        _msg_nh->setCallbackQueue(&_msg_callbacks);
+        ROS_DEBUG("Initialized all node handles");
+
+        _default_nh->param("spin_rate", _default_spin_rate, 10);
+        _default_nh->param("trajectory_publish_rate", _min_traj_publish_tickrate_hz, 10);
+
+        ROS_DEBUG_STREAM("Initalized params with default_spin_rate " << _default_spin_rate 
+            << " and trajectory_publish_rate " << _min_traj_publish_tickrate_hz);
+
+        this->_plan_sub = this->_msg_nh->subscribe<cav_msgs::TrajectoryPlan>("trajectory", 1000, &TrajectoryExecutor::onNewTrajectoryPlan, this);
```

Comment:
I think we can lower the queue size. Because there is a server delay, we do not want to publish a trajectory from few seconds ago. It should be a low risk to just drop it. Do you agree?

Reply:
Agreed